TypeSystem/Reflection/types.h

```diff
@@ -180,23 +181,35 @@ CURRENT_STRUCT(ReflectedType_Struct_Field) {
 
 CURRENT_STRUCT(ReflectedType_Struct, ReflectedTypeBase) {
   CURRENT_FIELD(native_name, std::string);
-  CURRENT_FIELD(super_id, TypeID, TypeID::UninitializedType);
+  CURRENT_FIELD(super_id, TypeID);
   CURRENT_FIELD(template_id, Optional<TypeID>);  // For instantiated `CURRENT_STRUCT_T`-s.
   CURRENT_FIELD(fields, std::vector<ReflectedType_Struct_Field>);
   CURRENT_DEFAULT_CONSTRUCTOR(ReflectedType_Struct) {}
-  // The `CanonicalName()` method serves two purposes:
-  // 1) it generates the unique name, for a template-instantiated `CURRENT_STRUCT_T` as well, and
-  // 2) it makes sure the the 'autogen -> #include -> autogen' loop does not change TypeIDs of templated types.
-  std::string CanonicalName() const {
+  // The `TemplateInnerTypeExpandedName()` method serves two purposes:
+  // 1) It generates the unique name for the structure for the schema-exported `CURRENT_STRUCT_T`-s.
+  //    (As `CURRENT_STRUCT_T`-s are exported as regular `CURRENT_STRUCT`-s, one per instantiation.)
+  // 2) It preserves the TypeID-s the way they used to be before the refactoring of the reflector.
+  std::string TemplateInnerTypeExpandedName() const {
     if (Exists(template_id)) {
-      // TODO(dkorolev): This code relies on `<>` at the end of the names of `CURRENT_STRUCT_T`-s. Fix it.
       CURRENT_ASSERT(native_name.length() > 2);
       CURRENT_ASSERT(native_name.substr(native_name.length() - 2) == "_Z");
       return native_name.substr(0, native_name.length() - 2) + "_T" + current::ToString(Value(template_id));
     } else {
       return native_name;
     }
   }
+  // TemplateFullNameDecorator() returns a function that converts names ending with "_Z"
+  // into the names ending with "_T...", where `...` is the TypeID of the `template_id` of this struct.
+  // Prerequisite: The struct is an instantiation of the template (or is declared so).
+  std::function<std::string(const std::string&)> TemplateFullNameDecorator() const {
+    CURRENT_ASSERT(Exists(template_id));
+    const TypeID t = Value(template_id);
+    return [t](const std::string& input) -> std::string {
+      CURRENT_ASSERT(input.length() >= 3);
+      CURRENT_ASSERT(input.substr(input.length() - 2) == "_Z");
+      return input.substr(0, input.length() - 1) + 'T' + current::ToString(static_cast<uint64_t>(t));
```

Comment:
Can this be a pure function that takes two arguments and is used both here and above?